### ISSUE TITLE
MonadIO/MonadBaseControl

### ISSUE THREAD

**Issue by pakoito** [kategory/kategory#474]
Taken from: http://www.parsonsmatt.org/2017/11/21/monadbasecontrol_in_five_minutes.html

**Comment by pakoito** [comment]
Stretch goal: `MonadBracket` for resource management https://www.fpcomplete.com/blog/2017/02/monadmask-vs-monadbracket

**Comment by 1Jajen1** [comment]
What exactly is needed for these two? Is it just a typeclass with a method each and instances for IO, mtl and the reactor/rx2 types? Because this would be quite useful in mtl style programs which use typeclasses and thus a polymorphic monad-stack . Currently you can do side-effects and almost all of `IO` in there by using `MonadDefer` and above, but you can't really interop with `IO` nicely.

**Comment by pakoito** [comment]
`MonadIO` exists based on the core idea that everything in Haskell should boil down to IO. Our primitive is actually `suspend`, except that it doesn't have cancellation baked so we have to stick to `IO` for now.

**Comment by 1Jajen1** [comment]
Yes I agree suspend makes a good marker, but as you said it does not have the same capabilities, so I'd like to have some sort of `MonadIO` so that existing methods returning `IO` can be included in mtl style monad-stacks.

**Comment by pakoito** [comment]
Yeah, both is good.

**Comment by 1Jajen1** [comment]
I'll see if I can do something there then :) `MonadIO` instances for mtl + IO and third party stuff should exists imo. There is also probably an easy implementation using the concurrent instances of those types, but we'll see :D
